src/runtime_src/xdp/profile/device/profile_ip_access.cpp

```diff
@@ -0,0 +1,179 @@
+    exclusive = false;
+    mapped = false;
+    mapped_address = 0;
+    device_handle = nullptr;
+    ip_index = -1;
+
+    // check for exclusive access to this IP
+    request_exclusive_ip_access(handle, index);
+    if (exclusive) {
+        device_handle = handle;
+        ip_index = index;
+    } else {
+        show_warning("Cannot get exclusive access");
+    }
+}
+
+ProfileIP::~ProfileIP() {
+    if (mapped) {
+        unmap();
+    }
+    if (exclusive) {
+        release_exclusive_ip_access(device_handle, ip_index);
+    }
+} 
+
+void ProfileIP::request_exclusive_ip_access(xclDeviceHandle handle, int index) {
+    /**
+     * TODO: when the XRT implements the exclusive context hal API, this
+     * method should try to open a exclusive context here and set the
+     * exclusive flag to be true if successful and save the context object
+     * into a member variable. If it fails, it should show a proper warning
+     * and return gracefully.
+     */
+    exclusive = true;
+    return;
+}
+
+void ProfileIP::release_exclusive_ip_access(xclDeviceHandle handle, int index) {
+    /**
+     * TODO: when the XRT implements the exclusive context hal API, this
+     * method should close the previously requested exclusive context if
+     * one was request and then set the exclusive flag back to false. If
+     * it fails, it should show a proper warning and return gracefully.
+     */
+    exclusive = false;
+    return;
+}
+
+void ProfileIP::map() {
+    /**
+     * TODO: so far we are asking the debug_ip_layout where the IP is. Once the XRT hal implements
+     * the function that maps the IP registers to user memory space, this method should be simplified
+     * to one function call to the hal API and saves the result in a mapped_address and set mapped flag.
+     */
+    if (!exclusive) {
+        return;
+    }
+    std::string subdev = "icap";
+    std::string entry = "debug_ip_layout";
+    size_t max_path_size = 256;
+    char raw_debug_ip_layout_path[max_path_size];
+    int get_sysfs_ret = xclGetSysfsPath(device_handle, subdev.c_str(), entry.c_str(), raw_debug_ip_layout_path, max_path_size);
+    if (get_sysfs_ret < 0) {
+        show_warning("Get debug_ip_layout path failed");
+        return;
+    }
+    std::string debug_ip_layout_path(raw_debug_ip_layout_path);
+    std::ifstream debug_ip_layout_fs(debug_ip_layout_path.c_str(), std::ifstream::binary);
+    size_t max_sysfs_size = 65536;
+    char buffer[max_sysfs_size];
+    if (debug_ip_layout_fs) {
+        debug_ip_layout_fs.read(buffer, max_sysfs_size);
+        if (debug_ip_layout_fs.gcount() > 0) {
+            debug_ip_layout* layout = (debug_ip_layout*)(buffer);
+            if (ip_index >= layout->m_count) {
+                show_warning("ip_index out of bound");
+                return;
+            }
+            debug_ip_data ip_data = layout->m_debug_ip_data[ip_index];
+            ip_name = std::string((char*)(&ip_data.m_name));

```

Comment:
This looks like a Coverity defect?   I suggest
ip_name.assign(reinterpret_cast<const char*>(&ip_data.m_name),64);
this will ensure that max 64 characters are read from a possible non-terminated char array.

Reply:
Changed to suggested approach in the new push.